synapse/push/httppusher.py

```diff
@@ -106,9 +119,118 @@ def __init__(self, hs: "HomeServer", pusher_config: PusherConfig):
+                    self.counts_only_type = self.data["counts_only_type"]
+                    del self.sanitized_data["counts_only_type"]
+
+                del self.sanitized_data["public_key"]
+
+    def _encrypt_notification_dict(
+        self, payload: Dict[str, Any], counts_only: bool = False
+    ) -> Dict[str, Any]:
+        """Called to process a payload according to the algorithm the pusher is
+        configured with. Namely, if the algorithm is `com.famedly.curve25519-aes-sha2`
+        we will encrypt the payload.
+
+        Args:
+            payload: The payload that should be processed
+        """
+        if (
+            self.hs.config.experimental.msc3013_enabled
+            and self.algorithm == "com.famedly.curve25519-aes-sha2"
+        ):
+            # we have an encrypted pusher, encrypt the payload
+            cleartext_notif = payload["notification"]
+            devices = cleartext_notif["devices"]
+            del cleartext_notif["devices"]
+            cleartext = json.dumps(cleartext_notif)
+
+            # create an ephemeral curve25519 keypair
+            private_key = X25519PrivateKey.generate()
+            # do ECDH
+            secret_key = private_key.exchange(self.public_key)
+            # expand with HKDF
+            zerosalt = bytes([0] * 32)
+            prk = hmac.new(zerosalt, secret_key, hashlib.sha256).digest()
+            aes_key = hmac.new(prk, bytes([1]), hashlib.sha256).digest()
+            mac_key = hmac.new(prk, aes_key + bytes([2]), hashlib.sha256).digest()
+            aes_iv = hmac.new(prk, mac_key + bytes([3]), hashlib.sha256).digest()[0:16]
+            # create the ciphertext with AES-CBC-256
+            encryptor = Cipher(algorithms.AES(aes_key), modes.CBC(aes_iv)).encryptor()
+            # AES blocksize is always 128 bits
+            padder = padding.PKCS7(128).padder()
+            ciphertext = (
+                encryptor.update(
+                    padder.update(cleartext.encode("utf-8")) + padder.finalize()
+                )
+                + encryptor.finalize()
+            )
+            # create the mac
+            mac = hmac.new(mac_key, ciphertext, hashlib.sha256).digest()[0:8]
+
+            d = {
+                "ephemeral": unpaddedbase64.encode_base64(
+                    private_key.public_key().public_bytes(
+                        Encoding.Raw, PublicFormat.Raw
+                    )
+                ),
+                "ciphertext": unpaddedbase64.encode_base64(ciphertext),
+                "mac": unpaddedbase64.encode_base64(mac),
+                "devices": devices,
+            }
+
+            # now, if the push frame is a counts only frame, we have to respect the counts_only_type setting
+            if counts_only:
+                if self.counts_only_type == "boolean":
+                    d["is_counts_only"] = True
+                elif self.counts_only_type == "full":
+                    d["counts"] = cleartext_notif["counts"]
```

Comment:
See the MSC https://github.com/Sorunome/matrix-doc/blob/soru/encrypted-push/proposals/3013-encrypted-push.md#ios

Why `curve25519-aes-sha2` specifically? Because only plain and curve25519-aes-sha2 exist in this PR, and it is *not* in plain

Reply:
so maybe call the method `_process_notification_dict_for_sending` instead?